File: WebStudents/src/Services/StudentService.cs
using Microsoft.EntityFrameworkCore;
using StudentsPerformance.Models;
using WebStudents.Dtos;
using WebStudents.src.EF;

namespace WebStudents.src.Services;

public class StudentService
{
    private readonly StudentDbContext _context;

    public StudentService(StudentDbContext context)
    {
        _context = context;
    }

    public IEnumerable<Student> GetAllStudents()
    {
        return _context.Student
            .Include(s => s.StudentGroup)
            .Include(s => s.Course)
            .OrderBy(s => s.LastName)
            .ThenBy(s => s.FirstName)
            .ToList();
    }

    public Student? GetStudentById(Guid id)
    {
        return _context.Student
            .Include(s => s.StudentGroup)
            .Include(s => s.Course)
            .FirstOrDefault(s => s.Id == id);
    }

    public void AddStudent(Student student)
    {
        _context.Student.Add(student);
        _context.SaveChanges();
    }

    public void UpdateStudent(Student student)
    {
        _context.Student.Update(student);
        _context.SaveChanges();
    }

    public void DeleteStudent(Guid id)
    {
        var student = _context.Student.Find(id);
        if (student != null)
        {
            _context.Student.Remove(student);
            _context.SaveChanges();
        }
    }

    public async Task<StudentSummaryDto?> GetStudentSummaryAsync(Guid studentId)
    {
        var student = await _context.Student.FirstOrDefaultAsync(s => s.Id == studentId);
        if (student == null) return null;

        var finals = await _context.FinalGrades
            .Where(f => f.StudentId == studentId)
            .Include(f => f.GradeSheet)
                .ThenInclude(s => s!.DisciplineOffering)
                    .ThenInclude(o => o!.Discipline)
            .Include(f => f.GradeSheet)
                .ThenInclude(s => s!.DisciplineOffering)
                    .ThenInclude(o => o!.Semester)
                        .ThenInclude(se => se!.AcademicYear)
            .Select(f => new FinalGradeSummaryItemDto
            {
                GradeSheetId = f.GradeSheetId,
                DisciplineName = f.GradeSheet!.DisciplineOffering!.Discipline!.Name,
                SemesterNumber = f.GradeSheet.DisciplineOffering!.Semester!.Number,
                AcademicYearStart = f.GradeSheet.DisciplineOffering.Semester!.AcademicYear!.StartYear,
                AcademicYearEnd = f.GradeSheet.DisciplineOffering.Semester!.AcademicYear!.EndYear,
                FinalScore = f.FinalScore,
                FinalMark = f.FinalMark,
                UpdatedAt = f.UpdatedAt
            })
            .OrderBy(x => x.AcademicYearStart)
            .ThenBy(x => x.SemesterNumber)
            .ToListAsync();

        return new StudentSummaryDto
        {
            StudentId = student.Id,
            FullName = $"{student.LastName} {student.FirstName}",
            Finals = finals
        };
    }
}

File: WebStudentsUI/src/app/services/student.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface Student {
  id?: string;
  firstName: string;
  lastName: string;
  gender: number;
  dateOfBirth: string;
  enrollmentDate: string;
  courseId?: number | null;
  studentGroupId?: string | null;
  studentGroup?: {
    id: string;
    name: string;
    studyYear: number;
  } | null;
}

export interface StudentFinalSummaryItem {
  gradeSheetId: string;
  disciplineName: string;
  semesterNumber: number;
  academicYearStart: number;
  academicYearEnd: number;
  finalScore?: number | null;
  finalMark?: string | null;
  updatedAt: string;
}

export interface StudentSummary {
  studentId: string;
  fullName: string;
  finals: StudentFinalSummaryItem[];
}

@Injectable({ providedIn: 'root' })
export class StudentService {
  private apiUrl = `${environment.apiUrl}/Student`;

  constructor(private http: HttpClient) {}

  getAll(): Observable<Student[]> {
    return this.http.get<Student[]>(this.apiUrl);
  }

  getById(id: string): Observable<Student> {
    return this.http.get<Student>(`${this.apiUrl}/${id}`);
  }

  create(student: Student): Observable<Student> {
    return this.http.post<Student>(this.apiUrl, student);
  }

  getMySummary(): Observable<StudentSummary> {
    return this.http.get<StudentSummary>(`${this.apiUrl}/me/summary`);
  }
}
